crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -0,0 +1,145 @@
+
+use super::module_resolver::PgModuleResolver;
+
+pub(crate) mod diesel_macro {
+    macro_rules! read_only_blocking {
+        ($pool:expr, $query:expr) => {{
+            let mut pg_pool_conn =
+                crate::context_data::db_data_provider::get_pg_pool_connection($pool)?;
+            pg_pool_conn
+                .build_transaction()
+                .read_only()
+                .run($query)
+                .map_err(|e| Error::Internal(e.to_string()))
+        }};
+    }
+    pub(crate) use read_only_blocking;
+}
+
+pub fn get_pg_pool_connection(pool: &PgConnectionPool) -> Result<PgPoolConnection, Error> {
+    pool.get().map_err(|e| {
+        Error::Internal(format!(
+            "Failed to get connection from PG connection pool with error: {:?}",
+            e
+        ))
+    })
+}
+
+pub fn establish_connection_pool() -> PgConnectionPool {
+    let database_url = env::var("PG_DB_URL").expect("PG_DB_URL must be set");
+    let manager = ConnectionManager::<PgConnection>::new(database_url);
+    r2d2::Pool::builder()
+        .build(manager)
+        .expect("Failed to create pool.")
+}
+
+#[allow(dead_code)]
+pub(crate) struct PgManager {
+    pub pool: PgConnectionPool,
+    pub module_cache: Arc<SyncModuleCache<PgModuleResolver>>,
+}
+
+impl PgManager {
+    pub(crate) fn new() -> Self {
+        let pool = establish_connection_pool();
+        Self {
+            pool: pool.clone(),
+            module_cache: Arc::new(SyncModuleCache::new(PgModuleResolver::new(pool))),
+        }
+    }
+
+    pub(crate) async fn fetch_tx(&self, digest: &str) -> Result<Option<StoredTransaction>, Error> {
+        let digest = Digest::from_str(digest)?;
+        read_only_blocking!(&self.pool, |conn| {
+            transactions::dsl::transactions
+                .filter(transactions::dsl::transaction_digest.eq(digest.into_vec()))
+                .get_result::<StoredTransaction>(conn) // Expect exactly 0 to 1 result
+                .optional()
+        })
+    }
+
+    pub(crate) async fn fetch_latest_epoch(&self) -> Result<StoredEpochInfo, Error> {
+        read_only_blocking!(&self.pool, |conn| {
+            epochs::dsl::epochs
+                .order_by(epochs::dsl::epoch.desc())
+                .limit(1)
+                .first::<StoredEpochInfo>(conn)
+        })
+    }
+
+    pub(crate) async fn fetch_epoch(
+        &self,
+        epoch_id: u64,
+    ) -> Result<Option<StoredEpochInfo>, Error> {
+        let bigint_e = i64::try_from(epoch_id)
+            .map_err(|_| Error::Internal("Failed to convert epoch id to i64".to_string()))?;
+
+        read_only_blocking!(&self.pool, |conn| {
+            epochs::dsl::epochs
+                .filter(epochs::dsl::epoch.eq(bigint_e))
+                .get_result::<StoredEpochInfo>(conn) // Expect exactly 0 to 1 result
+                .optional()
+        })
+    }
+
+    pub(crate) async fn fetch_epoch_strict(&self, epoch_id: u64) -> Result<StoredEpochInfo, Error> {
```

Comment:
Maybe a comment about the difference between this method and `fetch_epoch` could be useful.

Reply:
fetch_epoch returns Result<Option<StoredEpochInfo>> and _strict returns Result<StoredEpochInfo>, throwing if an epoch is not found. my thinking was that, on top-level queries, we can be more lenient with a data fetch from db, it's not uncommon for callers to provide an epoch_id that doesn't exist. But for field-level resolvers on TransactionBlock or Checkpoint, it would be a db issue if a checkpoint or tx exists but its associated epoh doesn't